### PR TITLE
pass code registry to all directive environments

### PR DESCRIPTION
**Pull request by koenpunt**
### :pencil: Description

To override datafetchers for all fields in an object using a directive, the environment should expose the `GraphQLCodeRegistry`.

Or at least the only way of overriding datafetchers from a directive on an object was by iterating over the fields and setting the datafetcher.

Something like this I expect to work:

```kt
override fun onObject(environment: KotlinSchemaDirectiveEnvironment<GraphQLObjectType>): GraphQLObjectType {
    val obj = environment.element
    val reg = environment.codeRegistry

    val expressionArgument = environment.directive.getArgument("expression")
    val expressionString = expressionArgument?.value as? String ?: return obj

    val authorizationManager = authorizationManagerForExpression(expressionString)

    obj.fieldDefinitions.forEach { field ->
        val fetcher = reg.getDataFetcher(obj, field)
        val authDataFetcher = AuthDataFetcher(fetcher, authorizationManager)
        reg.dataFetcher(obj, field, authDataFetcher)
    }

    return obj
}
```

Which I think should be possible with the provided change.

### CODE REVIEW

**Comment by dariuszkuc** [comment]
Hello 👋 
Can you elaborate on your use case? Why would you annotate the object vs the fields? i.e. given a schema (assuming there is `@authenticated` directive)

```
type Query {
  foo: Foo @authenticated
}

type Foo {
  bar: Int!
  baz: String!
}
```

In order to retrieve `Foo` you have to execute `foo` resolver which is already marked as `@authenticated`

**Comment by koenpunt** [comment]
For the use case of an `@auth` directive; if a type is used multiple times in the schema, then for every time the type is used the field should be annotated. With support for adding it to the object you don't have to;

<table>
<tr>
<td>

```graphql
type Query {
  other: OtherType
  foo: Foo @auth
}

type OtherType {
  foo: Foo @auth
}

type Foo {
  bar: Int!
  baz: String!
}
```

</td>
<td>VS</td>
<td>

```graphql
type Query {
  other: OtherType
  foo: Foo
}

type OtherType {
  foo: Foo
}

type Foo @auth {
  bar: Int!
  baz: String!
}
```

</td>
</tr>
</table>

Or if all the fields of a type require authorization, except one, you could annotate the object, and the field with the exception;

```graphql
type OtherType @auth {
  foo: Foo
  something: String
  other: OtherType @public
}
```

The latter example would probably require a less trivial implementation, but should still be doable.

The Node implementation of Apollo Server also supports a similar directive, as documented here: https://www.apollographql.com/docs/apollo-server/schema/creating-directives/#enforcing-access-permissions

And apart from authorization, this could also be used to map all fields through a specific datafetcher, for whatever purpose it might be necessary.

File: generator/graphql-kotlin-schema-generator/src/test/kotlin/com/expediagroup/graphql/generator/directives/KotlinDirectiveWiringFactoryTest.kt [hunk 1]
```diff
@@ -224,19 +224,19 @@ class KotlinDirectiveWiringFactoryTest {
...
-            .type { context, visitor -> context.thisNode().accept(context, visitor) }
-            .description("My Field Description")
-            .withDirective(graphQLLowercaseDirective)
-            .build()
-
-        assertFailsWith(InvalidSchemaDirectiveWiringException::class) {
-            SimpleWiringFactory().onWire(graphQLSchemaElement = myTestField, coordinates = mockk(), codeRegistry = null)
-        }
-    }
+//    @Test
```
Comment (dariuszkuc) [inline generator/graphql-kotlin-schema-generator/src/test/kotlin/com/expediagroup/graphql/generator/directives/KotlinDirectiveWiringFactoryTest.kt:227]: can you remove this commented out code

**Review by dariuszkuc — CHANGES_REQUESTED** [review]

File: generator/graphql-kotlin-schema-generator/src/test/kotlin/com/expediagroup/graphql/generator/directives/KotlinDirectiveWiringFactoryTest.kt (hunk 1 above)
Comment (koenpunt) [reply generator/graphql-kotlin-schema-generator/src/test/kotlin/com/expediagroup/graphql/generator/directives/KotlinDirectiveWiringFactoryTest.kt]: Done

File: generator/graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/directives/KotlinDirectiveWiringFactory.kt [hunk 2]
```diff
@@ -65,7 +65,8 @@ open class KotlinDirectiveWiringFactory(
             } else {
                 KotlinSchemaDirectiveEnvironment(
                     element = modifiedObject,
-                    directive = directive
+                    directive = directive,
+                    codeRegistry = codeRegistry ?: throw InvalidSchemaDirectiveWiringException("Unable to wire directive on an object due to a missing code registry")
```
[the remark this answers is not in the document]
Comment (koenpunt) [reply generator/graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/directives/KotlinDirectiveWiringFactory.kt]: Done

**Review by dariuszkuc — APPROVED** [review]

**Comment by dariuszkuc** [comment]
Bump to major as it is a breaking change to hook signature.

**Pull request merged**